### PR TITLE
Fix catapult miss animation: show this animation near the actual target object

### LINKED ISSUE #3127: Castle Siege, all missed shots of catapult are drawn beside balista.
During one siege my Catapult shot the main tower of castle (balista) for more than 10 times and it still was alive!

Edit:
every time catapult misses - we can watch an animation of stone beside the central tower (ballista)

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
fix #3127

### CODE REVIEW

**Comment by oleg-derevenetz** [comment]
Hi @Branikolog could you please take a look at this when you have some free time? I don't like that when animating the miss in OG style (if I understand correctly, in case of a miss, the boulder should fall in the courtyard near the target) sometimes you may get the impression that boulder hits creature standing on the cell on which the boulder falls. Maybe it's better to animate the miss with small explosion over the actual target?

File: src/fheroes2/battle/battle_action.cpp [hunk 1]
```diff
@@ -865,14 +865,20 @@ void Battle::Arena::ApplyActionCatapult( Command & cmd )
         u32 shots = cmd.GetValue();
 
         while ( shots-- ) {
-            u32 target = cmd.GetValue();
+            int target = cmd.GetValue();
             u32 damage = cmd.GetValue();
+            bool hit = cmd.GetValue() != 0;
```
Comment (ihhub) [inline src/fheroes2/battle/battle_action.cpp:870]: We could make these variables `const`.

File: src/fheroes2/battle/battle_arena.cpp [hunk 2]
```diff
@@ -626,8 +626,13 @@ void Battle::Arena::CatapultAction( void )
         while ( shots-- ) {
             int target = catapult->GetTarget( values );
             u32 damage = std::min( catapult->GetDamage(), values[target] );
-            cmd << target << damage;
-            values[target] -= damage;
+            bool hit = catapult->GetHitOrMiss();
```
Comment (ihhub) [inline src/fheroes2/battle/battle_arena.cpp:629]: We could make this variable `const`.

File: src/fheroes2/battle/battle_catapult.cpp [hunk 3]
```diff
@@ -145,16 +141,16 @@ int Battle::Catapult::GetTarget( const std::vector<u32> & values ) const
...
-        }
+        return Rand::Get( targets );
     }
 
     DEBUG_LOG( DBG_BATTLE, DBG_TRACE, "target not found.." );
 
     return 0;
 }
+
+bool Battle::Catapult::GetHitOrMiss() const
```
Comment (ihhub) [inline src/fheroes2/battle/battle_catapult.cpp:152]: The name of the method is not very clear. What's about call it "IsNextShotHit"? Not an ideal one, of course.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @oleg-derevenetz , I left few style related comments. Could you please take a look?

File: src/fheroes2/battle/battle_catapult.cpp (hunk 3 above)
Comment (oleg-derevenetz) [reply src/fheroes2/battle/battle_catapult.cpp:152]: Yes, it sounds better, thanks :)

**Comment by ihhub** [comment]
@oleg-derevenetz , thanks a lot for fixing this issue!

**Comment by Branikolog** [comment]
> Hi @Branikolog could you please take a look at this when you have some free time? I don't like that when animating the miss in OG style (if I understand correctly, in case of a miss, the boulder should fall in the courtyard near the target) sometimes you may get the impression that boulder hits creature standing on the cell on which the boulder falls. Maybe it's better to animate the miss with small explosion over the actual target?

Sounds like a good idea for future improvement. It seems that in heroes 3 developers got rid of current miss animation. I find it weird too. We need to discuss whether make miss animation over the actual target or, probably, in front bottom area of the target castle part.
Until then, I think, we should focus on more important issues. :)

**Pull request merged**